### ISSUE TITLE
Why the node of result png is blank?

### ISSUE THREAD

**Issue by lipingliang** [mingrammer/diagrams#229]
dear friends:
in my program, the node of result png is blank, why? help me, thanks!
![event_processing](https://user-images.githubusercontent.com/58203858/87238518-dab7a280-c435-11ea-805e-621e92616bc7.png)

the code:

from diagrams import Cluster, Diagram
from diagrams.aws.compute import ECS, EKS, Lambda
from diagrams.aws.database import Redshift
from diagrams.aws.integration import SQS
from diagrams.aws.storage import S3

with Diagram("Event Processing", show=True):
    source = EKS("k8s source")

    with Cluster("Event Flows"):
        with Cluster("Event Workers"):
            workers = [ECS("worker1"),
                       ECS("worker2"),
                       ECS("worker3")]

        queue = SQS("event queue")

        with Cluster("Processing"):
            handlers = [Lambda("proc1"),
                        Lambda("proc2"),
                        Lambda("proc3")]

    store = S3("events store")
    dw = Redshift("analytics")

    source >> workers >> queue >> handlers
    handlers >> store
    handlers >> dw

**Comment by gabriel-tessier** [comment]
@lipingliang 
Do you have error in the console?
Which version do you use?

I tried with last version 0.15 your code and it generate this:
![event_processing](https://user-images.githubusercontent.com/283765/87289463-a1fff200-c537-11ea-98d2-ae15d7a45a27.png)

Maybe try the example code and see if the icon display... 

https://diagrams.mingrammer.com/docs/getting-started/examples

If icons don't display try to reinstall.

**Comment by lipingliang** [comment]
> 
> 
> @lipingliang
> Do you have error in the console?
> Which version do you use?
> 
> I tried with last version 0.15 your code and it generate this:
> ![event_processing](https://user-images.githubusercontent.com/283765/87289463-a1fff200-c537-11ea-98d2-ae15d7a45a27.png)
> 
> Maybe try the example code and see if the icon display...
> 
> https://diagrams.mingrammer.com/docs/getting-started/examples
> 
> If icons don't display try to reinstall.

thanks!

**Comment by lipingliang** [comment]
> 
> 
> > @lipingliang
> > Do you have error in the console?
> > Which version do you use?
> > I tried with last version 0.15 your code and it generate this:
> > ![event_processing](https://user-images.githubusercontent.com/283765/87289463-a1fff200-c537-11ea-98d2-ae15d7a45a27.png)
> > Maybe try the example code and see if the icon display...
> > https://diagrams.mingrammer.com/docs/getting-started/examples
> > If icons don't display try to reinstall.
> 
> thanks!

thanks! I have reinstall 3 times, the problem remains.

**Comment by lipingliang** [comment]
> 
> 
> > > @lipingliang
> > > Do you have error in the console?
> > > Which version do you use?
> > > I tried with last version 0.15 your code and it generate this:
> > > ![event_processing](https://user-images.githubusercontent.com/283765/87289463-a1fff200-c537-11ea-98d2-ae15d7a45a27.png)
> > > Maybe try the example code and see if the icon display...
> > > https://diagrams.mingrammer.com/docs/getting-started/examples
> > > If icons don't display try to reinstall.
> > 
> > 
> > thanks!
> 
> thanks! I have reinstall 3 times, the problem remains.

My diagrams version is 0.16.0. In console, the error: can't open clustered_web_services.

**Comment by leandrodamascena** [comment]
@mingrammer I have some experience with bug reporting in an open source project and I believe I can create a decorator to detect errors in the project and facilitate the reporting of bugs and problems like this. I think in this case it could be useful.

What do you think of the idea? I can do a PR.

**Comment by gabriel-tessier** [comment]
@lipingliang 
Can you prevent copy paste the previous message and only post your reply it will make more easy to follow. Thanks!

Can you put in gist again your code well formatted. (In your first post, there's indentations problems).
I expect that you may have some indentations errors that may raise the error you provided.
Let's go step by step and try to figure out what is wrong with your installation.

Can you provide more information about your environment (OS: Windows Mac Linux .... Python version etc...)
How did you install Diagrams? which version of graphiz do you have?
This may help us to reproduce the problem.

**Comment by lipingliang** [comment]
@gabriel-tessier

**Comment by mingrammer** [comment]
@leandrodamascena Could you please give me some more details? Pseudo code would be greatly helpful!

**Comment by gabriel-tessier** [comment]
@lipingliang 

I'm sorry I can't reproduce your problem but it look like it's a problem with your system installation... and I don't have windows 7 to test.
Your initial problem with the pictures that don't display can be maybe related to this issue, if the resources folder can't be find:
https://github.com/mingrammer/diagrams/issues/163

Maybe somebody else with Windows and similar environment can reproduce the error and solve your problem.

**Comment by gabriel-tessier** [comment]
@lipingliang 

Actually you can build your diagrams using docker, for the installation follow the instruction here: https://github.com/mingrammer/diagrams/blob/master/DEVELOPMENT.md

Then you can put you file with the python code in the root of the source directory and run  `docker exec diagrams-1.0 python your_file.py` and it will generate your diagrams file.

Hope it help.

**Comment by seqizz** [comment]
Hi there :wave: 

I have a similar issue, OS is Linux/NixOS, installed version 0.19.1.

Following code:
```
#!/usr/bin/env python3
from diagrams import Diagram
from diagrams.generic.os import LinuxGeneral
from diagrams.generic.os import Centos

with Diagram("Grouped Workers", show=False, direction="TB"):
    Centos("lb") >> [LinuxGeneral("worker1"), LinuxGeneral("w2")] >> LinuxGeneral("test")

```

Generates this:

![grouped_workers](https://user-images.githubusercontent.com/307899/115835535-a9cdba80-a416-11eb-821c-aab73b0d5ad9.png)

I suspect this is related with something missing, but I can't see any warnings/errors anywhere. Dependencies supplied are:
- jinja2 "2.11.2"
- graphviz "0.14.1"

Can I help debugging further?

**Comment by gabriel-tessier** [comment]
Hi @seqizz, 

Is graphviz installed on your system? https://www.graphviz.org/download/#linux
If it's installed can you provide the version? Thanks

Can you consider using docker as described here:
 https://github.com/mingrammer/diagrams/blob/master/DEVELOPMENT.md
 
 Hope it help you debugging your problem.

**Comment by seqizz** [comment]
Hi,

Yes, graphviz version 2.42.2 is installed.

I was trying to solve this without Docker, since I was planning to package it for NixOS also. But apparently it's already [packaged](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/diagrams/default.nix). Trying from that derivation, no issues anymore. :tada: 

Sorry for the noise & thanks!

**Comment by ElArtista** [comment]
Having the same issue, by using an Arch Linux package from the AUR:
![web_service](https://user-images.githubusercontent.com/4622346/134463658-621edb29-6580-43d3-beb7-f57a7fff8d05.png)

No errors in the stdout/stderr, nothing at all.

**Comment by ElArtista** [comment]
Package was missing resources. Sent instructions to the maintainer to fix this!

**Comment by lostinplace** [comment]
what were the instructions to make this work?

**Comment by tommycarstensen** [comment]
Was a solution to this ever provided? I also get this error message:

    raise EnvironmentError("Global diagrams context not set up")      
OSError: Global diagrams context not set up

**Comment by yuri-scaranni** [comment]
I was having the same error here, running the code in pycharm. I just change to run in command line
'python mycode.py'

And the image was generated in the correct way. I believe something is missing in my venv in pycharm, but command line works fine.

**Comment by Ferroman** [comment]
Same issue here:
```
dot --version
dot - graphviz version 12.2.1 (20241206.2353)

Package      Version
------------ -------
cfgv         3.4.0
diagrams     0.24.3
distlib      0.3.9
filelock     3.17.0
graphviz     0.20.3
identify     2.6.8
```

![Image](https://github.com/user-attachments/assets/73fa73e3-cfd2-42c5-9ab5-c763f8cb8024)

No errors from the code
Reinstall, other installation types doesn't help much.

**Comment by Alfablos** [comment]
Hi everyone,
first thing first thank you for this tool, I haven't used it but now that I've discovered it I'll use it A LOT :D

This issue affects me as well.

I'm building my shell with
```nix
              packages = [
                (writeShellScriptBin "pycharm" "${tmux}/bin/tmux new -d '$HOME/.local/share/JetBrains/Toolbox/apps/pycharm/bin/pycharm .'")
                python
                pythonPackages.uv
                pythonPackages.ruff
                pythonPackages.diagrams
                pythonPackages.graphviz
                pythonPackages.scour
                (graphviz.override { inherit cairo; })
              ];
```

But I'm actually using uv in a virtualenv for testing.
Nodes images do not appear in the diagram.

I made a simple test to abstract away NixOS caveats when it comes to managing dependencies. I wrote this Dockerfile, which provides an environment with just graphviz and diagrams:
```Dockerfile
FROM python:3.12-alpine3.20      # Also tried 3.13.2

RUN apk add --update --no-cache graphviz

RUN pip install diagrams

WORKDIR /workspace

ENTRYPOINT [ "/usr/local/bin/python3" ]
```
And then
```shell
docker build --tag diagrams .
docker run --rm -v "$PWD:/workspace" diagrams diagram.py
```

The result is as expected, a diagram showing images.

Now, since I ran that in a container and the container used to run as root (no USER directive in the Dockerfile) I had a file in the CWD named web_service.png owned by root. When I ran `python3 diagram.py` OUTSIDE the container environment, only relying on what the Nix shell was providing I got the following error:

<details>

<summary>Python stack trace<summary>

```text
Traceback (most recent call last):
  File "/nix/store/f0lv3hm8ivi6hpqpxkyaahcgd6wkjvbr-python3.11-graphviz-0.20.3/lib/python3.11/site-packages/graphviz/backend/execute.py", line 88, in run_check
    proc.check_returncode()
  File "/nix/store/98jyyrwvy52nsvbzv9g9zvl9hdqw6sq3-python3-3.11.12/lib/python3.11/subprocess.py", line 502, in check_returncode
    raise CalledProcessError(self.returncode, self.args, self.stdout,
subprocess.CalledProcessError: Command '[PosixPath('/nix/store/wf6qw8hlh9g24kk0kwnxafs22fmsqk24-graphviz-12.2.1/bin/dot'), '-Kdot', '-Tpng', '-O', 'web_service']' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/antonio/Documents/mydiagrams/diagram.py", line 30, in <module>
    with Diagram("Web Service", show=False):
  File "/nix/store/0nqf987bwy3xssckna6bn45c08k6r70y-python3.11-diagrams-0.24.4/lib/python3.11/site-packages/diagrams/__init__.py", line 165, in __exit__
    self.render()
  File "/nix/store/0nqf987bwy3xssckna6bn45c08k6r70y-python3.11-diagrams-0.24.4/lib/python3.11/site-packages/diagrams/__init__.py", line 199, in render
    self.dot.render(format=self.outformat, view=self.show, quiet=True)
  File "/nix/store/f0lv3hm8ivi6hpqpxkyaahcgd6wkjvbr-python3.11-graphviz-0.20.3/lib/python3.11/site-packages/graphviz/_tools.py", line 171, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/f0lv3hm8ivi6hpqpxkyaahcgd6wkjvbr-python3.11-graphviz-0.20.3/lib/python3.11/site-packages/graphviz/rendering.py", line 122, in render
    rendered = self._render(*args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/f0lv3hm8ivi6hpqpxkyaahcgd6wkjvbr-python3.11-graphviz-0.20.3/lib/python3.11/site-packages/graphviz/_tools.py", line 171, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/f0lv3hm8ivi6hpqpxkyaahcgd6wkjvbr-python3.11-graphviz-0.20.3/lib/python3.11/site-packages/graphviz/backend/rendering.py", line 326, in render
    execute.run_check(cmd,
  File "/nix/store/f0lv3hm8ivi6hpqpxkyaahcgd6wkjvbr-python3.11-graphviz-0.20.3/lib/python3.11/site-packages/graphviz/backend/execute.py", line 90, in run_check
    raise CalledProcessError(*e.args)
graphviz.backend.execute.CalledProcessError: Command '[PosixPath('/nix/store/wf6qw8hlh9g24kk0kwnxafs22fmsqk24-graphviz-12.2.1/bin/dot'), '-Kdot', '-Tpng', '-O', 'web_service']' returned non-zero exit status 1. [stderr: b'Warning: No such file or directory while opening /nix/store/0nqf987bwy3xssckna6bn45c08k6r70y-python3.11-diagrams-0.24.4/lib/python3.11/site-packages/resources/aws/network/elastic-load-balancing.png\nWarning: No or improper image="/nix/store/0nqf987bwy3xssckna6bn45c08k6r70y-python3.11-diagrams-0.24.4/lib/python3.11/site-packages/resources/aws/network/elastic-load-balancing.png" for node "9a3342f632fa43f59ecc5ebde784f583"\nWarning: No such file or directory while opening /nix/store/0nqf987bwy3xssckna6bn45c08k6r70y-python3.11-diagrams-0.24.4/lib/python3.11/site-packages/resources/aws/compute/ec2.png\nWarning: No or improper image="/nix/store/0nqf987bwy3xssckna6bn45c08k6r70y-python3.11-diagrams-0.24.4/lib/python3.11/site-packages/resources/aws/compute/ec2.png" for node "7f8c8d09a9c24405a078c66551f58253"\nWarning: No such file or directory while opening /nix/store/0nqf987bwy3xssckna6bn45c08k6r70y-python3.11-diagrams-0.24.4/lib/python3.11/site-packages/resources/aws/database/rds.png\nWarning: No or improper image="/nix/store/0nqf987bwy3xssckna6bn45c08k6r70y-python3.11-diagrams-0.24.4/lib/python3.11/site-packages/resources/aws/database/rds.png" for node "9a398b7744c2456082bc91fa31b3e122"\nError: Could not open "web_service.png" for writing : Permission denied\n']
```

</details>

TLDR: `dot` tried to convert the temporary file `web_service` (no extension) to `web_service.png` but was unable to find png files related to AWS resources. ALSO it tried to replace `web_service.png` with the new content but was unable to since the existing file was, luckily, owned by root.

I was left with a raw `web_service` file, with a well-formed container-generated web_service.png and the doubt that some files from the repo are not contained in the output of the nix derivation.

```shell
❯ ls /nix/store/0nqf987bwy3xssckna6bn45c08k6r70y-python3.11-diagrams-0.24.4/lib/python3.11/site-packages/resources/aws/compute/ec2.png
"/nix/store/0nqf987bwy3xssckna6bn45c08k6r70y-python3.11-diagrams-0.24.4/lib/python3.11/site-packages/resources/aws/compute/ec2.png": No such file or directory (os error 2)
```

As a matter of fact the content of `/nix/store/0nqf987bwy3xssckna6bn45c08k6r70y-python3.11-diagrams-0.24.4/lib/python3.11/site-packages` is `/nix/store/0nqf987bwy3xssckna6bn45c08k6r70y-python3.11-diagrams-0.24.4/lib/python3.11/site-packages/diagrams`, which contains:
```shell
.r--r--r-- 19k root  1 gen  1970  __init__.py
dr-xr-xr-x   - root  1 gen  1970  __pycache__
dr-xr-xr-x   - root  1 gen  1970  alibabacloud
dr-xr-xr-x   - root  1 gen  1970  aws
dr-xr-xr-x   - root  1 gen  1970  azure
dr-xr-xr-x   - root  1 gen  1970  base
dr-xr-xr-x   - root  1 gen  1970  c4
dr-xr-xr-x   - root  1 gen  1970  custom
dr-xr-xr-x   - root  1 gen  1970  digitalocean
dr-xr-xr-x   - root  1 gen  1970  elastic
dr-xr-xr-x   - root  1 gen  1970  firebase
dr-xr-xr-x   - root  1 gen  1970  gcp
dr-xr-xr-x   - root  1 gen  1970  generic
dr-xr-xr-x   - root  1 gen  1970  gis
dr-xr-xr-x   - root  1 gen  1970  ibm
dr-xr-xr-x   - root  1 gen  1970  k8s
dr-xr-xr-x   - root  1 gen  1970  oci
dr-xr-xr-x   - root  1 gen  1970  onprem
dr-xr-xr-x   - root  1 gen  1970  openstack
dr-xr-xr-x   - root  1 gen  1970  outscale
dr-xr-xr-x   - root  1 gen  1970  programming
dr-xr-xr-x   - root  1 gen  1970  saas
```

Each of those directories only contains python code:
```shell
❯ ls /nix/store/0nqf987bwy3xssckna6bn45c08k6r70y-python3.11-diagrams-0.24.4/lib/python3.11/site-packages/diagrams/alibabacloud
.r--r--r--  299 root  1 gen  1970  __init__.py
dr-xr-xr-x    - root  1 gen  1970  __pycache__
.r--r--r--  546 root  1 gen  1970  analytics.py
.r--r--r-- 1,3k root  1 gen  1970  application.py
.r--r--r--  360 root  1 gen  1970  communication.py
.r--r--r-- 1,7k root  1 gen  1970  compute.py
.r--r--r-- 1,7k root  1 gen  1970  database.py
.r--r--r--  475 root  1 gen  1970  iot.py
.r--r--r--  923 root  1 gen  1970  network.py
.r--r--r-- 1,6k root  1 gen  1970  security.py
.r--r--r--  948 root  1 gen  1970  storage.py
.r--r--r--  286 root  1 gen  1970  web.py
```
I'm trying to write an overlay but I s*ck at overlays :D

Note that **`dot -Kdot -Tpng  -O web_service` exits with because those files aren't found, so I think the code (which deletes the raw file anyway and suppresses the stderr) should instead propagate the error. 

From a (very) quick look it seems that the `render` method of the `Diagram` class is actually firing graphviz without handling the [graphviz.CalledProcessError](https://graphviz.readthedocs.io/en/stable/api.html#graphviz.CalledProcessError), which should blow up in my face. For some reason this doesn't happen when the file can be created but images are not found.

```python
    def render(self) -> None:
        if isinstance(self.outformat, list):
            for one_format in self.outformat:
                self.dot.render(format=one_format, view=self.show, quiet=True)
        else:
            self.dot.render(format=self.outformat, view=self.show, quiet=True)

```

The temporary solution for me is this (`flake.nix`):
```nix
            pkgs.mkShell {
              packages = [
                python
                pythonPackages.uv
                pythonPackages.ruff
                # pythonPackages.diagrams
                graphviz
                docker-client
              ];

              shellHook = ''
                export UV_PYTHON_DOWNLOADS=never

                source .venv/bin/activate
                exec -l zsh
              '';
            };

````

To start working from scratch:
```shell
nix develop
```

And in the shell:
```shell
uv init
```

`pyproject.toml`
```toml
[project]
name = "mydiagrams"
version = "0.1.0"
description = "Add your description here"
readme = "README.md"
requires-python = ">=3.11"
dependencies = [
    "diagrams==0.24.1",
]
```
```shell
uv venv  --python 3.13
uv sync

python diagram.py     # Generates the image correctly since it uses the virtualenv dependency, not the one in the nix store
```
TL;DR I think that not being able to fetch AWS pngs is probably due to wrong packagnig in nixpkgs on Nix/NixOS (@addict3d can you confirm?) rather than an issue with the code itself.
